sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -334,6 +340,160 @@ private case class ParquetFileInfoWithBlockMeta(filePath: Path, blocks: Seq[Bloc
     partValues: InternalRow, schema: MessageType, isCorrectedInt96RebaseMode: Boolean,
     isCorrectedRebaseMode: Boolean, hasInt96Timestamps: Boolean)
 
+class HMBSeekableInputStream(hmb: HostMemoryBuffer, hmbLength: Long) extends SeekableInputStream {
+  private var pos = 0L
+  private var mark = -1L
+  private val temp = new Array[Byte](8192)
+
+  override def read(): Int = {
+    if (pos >= hmbLength) {
+      -1
+    } else {
+      val result = hmb.getByte(pos)
+      pos += 1
+      result & 0xFF
+    }
+  }
+
+  override def read(buffer: Array[Byte], offset: Int, length: Int): Int = {
+    if (pos >= hmbLength) {
+      -1
+    } else {
+      val numBytes = Math.min(available(), length)
+      hmb.getBytes(buffer, offset, pos, numBytes)
+      pos += numBytes
+      numBytes
+    }
+  }
+
+  override def skip(count: Long): Long = {
+    val oldPos = pos
+    pos = Math.min(pos + count, hmbLength)
+    pos - oldPos
+  }
+
+  override def available(): Int = Math.min(hmbLength - pos, Integer.MAX_VALUE).toInt
+
+  override def mark(ignored: Int): Unit = {
+    mark = pos
+  }
+
+  override def reset(): Unit = {
+    if (mark <= 0) {
+      throw new IOException("reset called before mark")
+    }
+    pos = mark
+  }
+
+  override def markSupported(): Boolean = true
+
+  def getPos: Long = pos
+
+  override def seek(offset: Long): Unit = {
+    pos = offset
+  }
+
+  @throws[IOException]
+  override def readFully(buffer: Array[Byte]): Unit = {
+    val amountRead = read(buffer)
+    val remaining = buffer.length - amountRead
+    if (remaining > 0) {
+      throw new EOFException("Reached the end of stream with " + remaining + " bytes left to read")
+    }
+  }
+
+  @throws[IOException]
+  override def readFully(buffer: Array[Byte], offset: Int, length: Int): Unit = {
+    val amountRead = read(buffer, offset, length)
+    val remaining = length - amountRead
+    if (remaining > 0) {
+      throw new EOFException("Reached the end of stream with " + remaining + " bytes left to read")
+    }
+  }
+
+  @throws[IOException]
+  override def read(buf: ByteBuffer): Int =
+    if (buf.hasArray) {
```

Comment:
maybe a small comment somewhere on why we need two implementations one for direct and one for JVM byte buffers.